chapter/methodology.md

```diff
@@ -0,0 +1,964 @@
+    inner [ label = "Record" ]
+    file_main_container [label="RecordField| main_containers"]
+
+
+
+    lin:f1 -> outer
+    outer -> lin:f2 [style = dashed]
+    outer -> lin:f3 [style = dashed]
+    outer -> lin:f4 [style = dashed]
+
+    lin:f2 -> field_apiVersion
+    field_apiVersion -> lin:f5 [style = dashed]
+
+    lin:f6 -> inner
+    inner -> lin:f7 [style = dashed]
+
+    lin:f3 -> field_containers
+    field_containers -> lin:f6 [style = dashed]
+
+    lin:f4 -> field_Metadata
+    field_Metadata-> lin:f8 [style = dashed]
+
+    lin:f7 -> file_main_container
+    file_main_container -> lin:f8 [style = dashed]
+}
+```
+
+To provide the necessary references, NLS makes used of the *scope safe* memory of its `Linearizer` implementation.
+This is possible, because each record value corresponds to its own scope.
+The complete process looks as follows:
+
+1. When registering a record, first the outer `Record` is added to the linearization
+2. This is followed by `RecordField` items for its fields, which at this point do not reference any value.
+3. NLS then stores the `id` of the parent as well as the fields and the offsets of the corresponding items (`n-4` and `[(apiVersion, n-3), (containers, n-2), (metadata, n-1)]` respectively in the example [@fig:nls-lin-records]).
+4. The `scope` method will be called in the same order as the record fields appear.
+   Using this fact, the `scope` method moves the data stored for the next evaluated field into the freshly generated `Linearizer`
+5. **(In the sub-scope)** The `Linearizer` associates the `RecordField` item with the (now known) `id` of the field's value.
+   The cached field data is invalidated such that this process only happens once for each field.
+
+
+##### Variable Reference
+
+The usage of a variable is always expressed as a `Var` node that holds an identifier.
+Registering a name usage is a multi-step process.
+
+First, NLS tries to find the identifier in its scope-aware name registry.
+If the registry does not contain the identifier, NLS will linearize the node as `Unbound`.
+In the case that the registry lookup succeeds, NLS retrieves the referenced `Declaration` or `RecordField`. The linearizer will then add a usage item in the `Resolved` state to the linearization and update the declaration's list of usages.
+
+##### Resolution of Record Fields
+
+The AST representation of record destructuring in [@fig:nickel-static-access] shows that accessing inner records involves chains of unary operations *ending* with a reference to a variable binding.
+Each operation encodes one field of a referenced record.
+However, to reference the corresponding declaration, the final usage has to be known.
+Therefore, instead of linearizing the intermediate elements directly, the `Linearizer` adds them to a shared stack until the grounding variable reference is registered.
+
+Whenever a variable usage is linearized, NLS checks the stack for latent destructors.
```

Comment:
I think destructors haven't been used until now. It could be better for the reader to either define them quickly: 

```suggestion
Whenever a variable usage is linearized, NLS checks the stack for latent destructors (record accesses).
```

Or to stick to using `access` everywhere.

Reply:
It's your call, but I think it doesn't cost much to recall this here, as the reader may have forgotten, or dive directly into this chapter.